Upated pull request for closure plots functions after rebase

C-Wirth opened the pull request:
# Additions, Renames and Deletions
## Files added:
    include/universal/utility/generateClosurePlots.hpp
    playground/closure_of_8bit_systems.cpp

## Files renamed
    viewing.ipynb to draw_closure_plots.ipynb

## Files deleted
    playground/gemerateClosurePlots.cpp

# Notable changes
## 1.
The functions of generateClosurePlots.cpp are now in 
generateClosurePlots.hpp to best align with the header only paradigm of the library
To use generate the csv/txt files for a closure plot, configure and run 
generate_a_closure_map.cpp

## 2. 
draw_closure_plots.ipynb (previously known as viewing.ipynb) has been 
refactored, seriously.

## 3.
    I have not dove into lns systems yet so there is no progress with 
closure plots for lns, however closure plots for cfloats and posits can be 
generated from the last cell in the draw_closure_plots.ipynb file. Maybe 
the lns can be done too, this just has not been tested.

## 4.
closure_of_8_bit_systems.cpp generates the .txt/.csv files for the closure 
plots of 3 comparable number systems

Ravenwater commented on include/universal/utility/error.hpp:
```diff
@@ -35,5 +36,41 @@ ReturnType RelativeError(const ArgumentType& actual, const ArgumentType& referen
+/// <typeparam name="ReturnType">type representation of relative error. defaults to double</typeparam>
+
+/// <returns>relative error between actual and reference values</returns>
+template<typename ReturnType = double>
```
the template parameter ArgumentType has been removed erroneously. The design of this template function makes the argument types and the return type programmable to the context. Please add ArgumentType back into this function definition.

Ravenwater commented on include/universal/utility/generateClosurePlots.hpp:
```diff
@@ -0,0 +1,450 @@
+#ifndef GENERATECLOSUREPLOTS_H
```
replace the compilation guards with a simple
```
#pragma 
```

it is a more modern way to accomplish single inclusion.

Ravenwater commented on include/universal/utility/generateClosurePlots.hpp:
```diff
@@ -0,0 +1,450 @@
+ 
+ #include <universal/math/stub/classify.hpp>
+ #include <universal/utility/error.hpp>
+ #include <universal/utility/generateClosurePlots.hpp> 
```
shouldn't include yourself

Ravenwater commented on include/universal/utility/generateClosurePlots.hpp:
```diff
@@ -0,0 +1,450 @@
+     unsigned long overflow;
+     unsigned long underflow;
+     unsigned long saturate;
+     unsigned long absoluteError;
```
absolute and relative error are floating-point values.

Ravenwater commented on include/universal/utility/generateClosurePlots.hpp:
```diff
@@ -0,0 +1,450 @@
+         }
+ 
+         else{
+             absErr = std::abs(sw::universal::AbsoluteError(vcDouble, targetVal));
```
since you define this function in the namespace sw::universal, you do not need to explicitely reference them through a fully articulated name. Simply calling AbsoluteError() will do.

Ravenwater commented on include/universal/utility/generateClosurePlots.hpp:
```diff
@@ -0,0 +1,450 @@
+ }
+ 
+ template<typename NumberType>
+ void configure_value(NumberType v, std::string& vString){
```
you are switching type name structure. Previous names were camelCase, and now this one uses an underscore. I suggest you make this camelCase as well.

Ravenwater commented on playground/closure_of_8bit_systems.cpp:
```diff
@@ -0,0 +1,66 @@
+
+#include <universal/utility/generateClosurePlots.hpp>
+
+/// Version 13 Apirl 2025
```
spelling error of April

Ravenwater commented on playground/generate_a_closure_map.cpp:
```diff
@@ -0,0 +1,55 @@
+constexpr unsigned es{ 4 };
+using Real= posit<nbits, es>; 
+std::string type = "posit";
+///////////////////////////////     DONT EDIT BELOE THIS LINE       ///////////////////////////////
```
spelling error of Below

Ravenwater commented on playground/generate_a_closure_map.cpp:
```diff
@@ -0,0 +1,55 @@
+
+///////////////////////////////     CONFIGURE Real NUMBER SYSTEM BELOW       ///////////////////////////////
+namespace sw::universal{
+constexpr unsigned nbits{ 2 };
```
this is probably not what you want. You are injecting a constexpr symbol with a generic name into the sw::universal namespace. This could change all kinds of code. 

This configuration should move to inside main() and stay out of the sw::universal namespace.

like this:
```
int main()
   using namespace sw::universal;

   constexpr unsigned NBITS = 8; 
   constexpr unsigned ES = 4;                  // capitalized to indicate these are CONSTANTs
   using Real = posit<NBITS, ES>;
...
```

C-Wirth replied:
How can I pass 'Real' as the template in the explicit instantiation of buildClosurePlot<>() without declaring the Real, NBITS, and ES outside of main?

Ravenwater reviewed (changes requested):
minor changes requested. Looks good from a design point of view

C-Wirth commented:
I will push the changes to the pull request once we resolve my one concern wrt to the explicit instantiation of buildClosurePlot (in the comment above).  Other than that, all other issues have been resolved.

C-Wirth commented:
Also, from Grok I got that I can resolve the errors then push directly to the branch to update my pull request:

git add .
git commit -m "Addressed review comments"
git push origin stillwater-sc:v3.81

Is this the proper way to do so?

Ravenwater commented:
Try it and let me know. I typically update my own fork and that appears to
trickle down to the PR automatically.

   *Dr. E. Theodore L. Omtzigt*

CEO and Founder

Stillwater Supercomputing, Inc.



Mobile phone #  : (916) 296-7901

95 South Freeport Road, #720
South Freeport, ME 04078-0720



On Mon, Apr 14, 2025 at 11:01 PM C-Wirth ***@***.***> wrote:

> Also, from Grok I got that I can resolve the errors then push directly to
> the branch to update my pull request:
>
> git add .
> git commit -m "Addressed review comments"
> git push origin stillwater-sc:v3.81
>
> Is this the proper way to do so?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stillwater-sc/universal/pull/451#issuecomment-2803628499>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADR5UKNH32HQXR7HLDTGC32ZRY7HAVCNFSM6AAAAAB3DDD2IWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBTGYZDQNBZHE>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
> *C-Wirth* left a comment (stillwater-sc/universal#451)
> <https://github.com/stillwater-sc/universal/pull/451#issuecomment-2803628499>
>
> Also, from Grok I got that I can resolve the errors then push directly to
> the branch to update my pull request:
>
> git add .
> git commit -m "Addressed review comments"
> git push origin stillwater-sc:v3.81
>
> Is this the proper way to do so?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stillwater-sc/universal/pull/451#issuecomment-2803628499>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADR5UKNH32HQXR7HLDTGC32ZRY7HAVCNFSM6AAAAAB3DDD2IWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMBTGYZDQNBZHE>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

Ravenwater commented on include/universal/utility/error.hpp:
```diff
@@ -62,7 +62,7 @@ ReturnType LogRelativeError(const ArgumentType& actual, const ArgumentType& refe
 /// <returns>relative error between actual and reference values</returns>
-template<typename ReturnType = double>
+template<typename ArgumentType = double, typename ReturnType = double>
 ReturnType MinMaxLogNormalization(const ReturnType errVal, const ReturnType maxpos, const ReturnType minpos){ 
```
errVal, maxpos, minpos are all ArgumentType types, so they are incorrectly typed.

Ravenwater reviewed (approved):
Thank you
